source/Halibut.Tests/Transport/Streams/TimeoutStreamWrapperFixture.cs

```diff
@@ -0,0 +1,369 @@
+﻿using System;
+using System.Diagnostics;
+using System.IO;
+using System.Net;
+using System.Net.Sockets;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using FluentAssertions;
+using Halibut.Tests.Support;
+using Halibut.Tests.Support.TestAttributes;
+using Halibut.Transport.Streams;
+using NUnit.Framework;
+
+namespace Halibut.Tests.Transport.Streams
+{
+    [TestTimeout(timeoutInSeconds: 20)]
+    public class TimeoutStreamWrapperFixture : BaseTest
+    {
+        [Test]
+        public async Task ReadShouldPassThrough()
+        {
+            string? readData = null;
+            
+            var (disposables, sut, performListenerWrite) = await BuildTcpClientAndTcpListener(CancellationToken);
+
+            using (disposables)
+            {
+                await performListenerWrite("Test");
+
+                var buffer = new byte[19];
+                var readBytes = sut.Read(buffer, 0, 19);
+                if (readBytes > 0)
+                {
+                    readData = Encoding.UTF8.GetString(buffer, 0, readBytes);
+                }
+
+                Assert.AreEqual("Test", readData);
+            }
+        }
+        
+        [Test]
+        public async Task ReadAsyncShouldPassThrough()
+        {
+            string? readData = null;
+            
+            var (disposables, sut, performListenerWrite) = await BuildTcpClientAndTcpListener(CancellationToken);
+
+            using (disposables)
+            {
+                await performListenerWrite("Test");
+
+                var buffer = new byte[19];
+                var readBytes = await sut.ReadAsync(buffer, 0, 19, CancellationToken);
+                if (readBytes > 0)
+                {
+                    readData = Encoding.UTF8.GetString(buffer, 0, readBytes);
+                }
+
+                Assert.AreEqual("Test", readData);
+            }
+        }
+
+        [Test]
+        public async Task ReadAsyncShouldCancel()
+        {
+            using (var readTokenSource = new CancellationTokenSource(TimeSpan.FromSeconds(5)))
+            {
+                var (disposables, sut, _) = await BuildTcpClientAndTcpListener(
+                    CancellationToken,
+                    onListenerRead: async _ => await Task.Delay(TimeSpan.FromDays(1), CancellationToken));
+
+                using (disposables)
+                {
+                    // Ensure the timeouts are not used
+                    sut.WriteTimeout = (int)TimeSpan.FromSeconds(120).TotalMilliseconds;
+                    sut.ReadTimeout = (int)TimeSpan.FromSeconds(120).TotalMilliseconds;
+                    
+                    var stopWatch = Stopwatch.StartNew();
+                    Exception? actualException = null;
+
+                    // Brute force attempt to get the Write to be slow
```

Comment:
This comment is used below for both writes and reads.

In this case, it looks like it's to make 'Reads' slow?

Reply:
Good spot, the reads should not have been wrapped in the while loop as we can predictably delay read (we just don't write anything). The writes are the bit we are struggling to predictably delay